### ISSUE TITLE
Should implementers always define controller actions?

### ISSUE THREAD

**Issue by yellow5** [tiagopog/jsonapi-utils#28]
_This is a question/confirmation issue and not a bug/code flaw issue._

When using JSONAPI::Utils, are the controller actions expected to explicitly be defined?

When mixing in JSONAPI::Utils, [we also get the JSONAPI::ActsAsResourceController module](https://github.com/b2beauty/jsonapi-utils/blob/master/lib/jsonapi/utils.rb#L13) and its behavior. [Within that module](https://github.com/cerebris/jsonapi-resources/blob/master/lib/jsonapi/acts_as_resource_controller.rb), default actions are defined.

The first sentence in the ["How does it work?" section of the README](https://github.com/b2beauty/jsonapi-utils#how-does-it-work) made me think more about it:

> One of the main motivations behind JSONAPI::Utils is to keep things explicit in controllers so that developers can easily understand and maintain code.

With a primary motivation being explicit controller behavior, does that in turn mean that explicit controller actions are implied?

**Context**

While test driving some resources using JSONAPI::Utils + JSONAPI::Resources, I noticed that my immutable endpoints were working without defining any actions. At the time, I did not know why, and the specs were green so I let it be.

Recently, I encountered some trouble attempting to add a fully mutable resource, in particular with PATCH requests. After verifying that the API request body was correct, I narrowed the issue down to double parsing of the request ([once in JSONAPI::Utils](https://github.com/b2beauty/jsonapi-utils/blob/master/lib/jsonapi/utils/request.rb#L10-L16) and a [second time](https://github.com/cerebris/jsonapi-resources/blob/master/lib/jsonapi/acts_as_resource_controller.rb#L66-L68) in the default action definition in JSONAPI::Resources).

The double parsing happened since it passed through a before filter than used the RequestParser instance and then entered the default update action which initialized a second RequestParser instance.

While looking closer at the specs for JSONAPI::Utils, I noticed that the controllers were always defining the actions. This led me back to the README, since I figured that I must have missed something.

To sum it up, I think that I have been doing it wrong (from the JSONAPI::Utils perspective) and I am looking for confirmation. 😄

**Comment by tiagopog** [comment]
`JSONAPI::Utils` takes only advantage of the `JSONAPI::Resources`' resource layer for proper serialization and in order to use some of its APIs to validate requests against the resource definitions.

With `JSONAPI::Utils` it's recommended to write your actions and its inner operations (service objects, interactors etc) explicitly – rather than taking them into resources – so that you improve the expressiveness and maintenance of your code.

Well, having this issue in mind, we could think in a way to avoid the double parsing of the `JSONAPI::RequestParser` object when it deals with default actions.

Thoughts?

**Comment by yellow5** [comment]
Trying to prevent the double parsing definitely has some challenges. `JSONAPI::Utils` has to consider how forceful to participate in preventing the risk with how tightly to couple with the `JSONAPI::Resources` library. Hopefully this can help you determine a path forward. 😄 

Here are some options that take different approaches with `JSONAPI::Utils` lib's involvement with prevention during implementation:

1. Documentation informs risk
2. Warning log entries when using defaults
3. Override default actions and raise errors
4. Override `JSONAPI::Resources` #process_request method

Here's what I see as the positives/negatives of each (just my two cents):

**1. Documentation informs risk**

👍 

* No risk of accidentally disrupting any downstream code
* No pressure on `JSONAPI::Utils` to keep a watchful eye on JR for changes
* Documentation can change without modifying any library behavior

👎 

* Double parsing still happens

**2. Warning log entries when using defaults**

👍 

* Actively helps expose the risk during runtime
* Gets overridden when Future Developer defines their own actions

👎 

* Double parsing still happens
* `JSONAPI::Utils` has to alias and call JR actions

**3. Override default actions and raise errors**

👍 

* Guarantees that the risk will not happen

👎 

* Forces Future Developer to define their own actions (no longer encouraged)

**4. Override `JSONAPI::Resources` #process_request method**

👍 

* Guarantees that the risk will not happen

👎 

* Pressure on `JSONAPI::Utils` to tightly couple parsing method to `JSONAPI::Resources` method name

**Comment by tiagopog** [comment]
Good list, @yellow5! As the gem's approach is to stay as a layer on top of JR, it seems reasonable to go for a combination of the points **1** and **2** :-)
